rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,308 @@
+        submitForm(data)
+          .then(() => {})
+          /**
+           * Map errors from your server response into a structure you'd like to work with.
+           * In this case resulting in this object: `{ name: false, email: true, age: true }`
+           */
+          .catch((errors) => setServerErrors(mapServerErrors(errors)));
+      }}
+      onClearServerErrors={() => setServerErrors({ name: false, email: false, age: false })}
+    >
+      <Form.Field name="name" serverInvalid={serverErrors.name}>
+        <Form.Label>Full name</Form.Label>
+        <Form.Control />
+        <Form.Message match="missingValue" forceMatch={serverErrors.name}>
+          Please enter your name.
+        </Form.Message>
+      </Form.Field>
+
+      <Form.Field name="email" serverInvalid={serverErrors.email}>
+        <Form.Label>Email address</Form.Label>
+        <Form.Control type="email" />
+        <Form.Message match="missingValue">Please enter your email.</Form.Message>
+        <Form.Message match="typeMismatch" forceMatch={serverErrors.email}>
+          Please provide a valid email.
+        </Form.Message>
+      </Form.Field>
+
+      <Form.Field name="age" serverInvalid={serverErrors.age}>
+        <Form.Label>Age</Form.Label>
+        <Form.Control type="number" min="18" max="57" step="1" />
+        <Form.Message match="missingValue">Please enter your age.</Form.Message>
+        <Form.Message match="rangeOverflow">You must be 18 or older.</Form.Message>
+        <Form.Message match="rangeUnderflow">You must be 57 or younger.</Form.Message>
+        <Form.Message match="stepMismatch">Please enter a whole number.</Form.Message>
+        {serverErrors.age ? (
+          <Form.Message>Registration not open to people aged 40 exactly.</Form.Message>
+        ) : null}
+      </Form.Field>
+
+      <Form.Submit>Submit</Form.Submit>
+    </Form.Root>
+  );
+}
+```
+
+You should also clear the server errors using the `onClearServerErrors` callback prop on the `Form.Root` part.
+
+> This will clear the server errors before the form is re-submitted, and when the form is reset.
+
+On top of this, it means you have control over when to reset single server errors.
+For example you could reset the email server error as soon as the user edits it:
+
+```jsx
+<Form.Field name="email" serverInvalid={serverErrors.email}>
+  <Form.Label>Email address</Form.Label>
+  <Form.Control
+    type="email"
+    onChange={() => setServerErrors((prev) => ({ ...prev, email: false }))}
+  />
+  <Form.Message match="missingValue">Please enter your email.</Form.Message>
+  <Form.Message match="typeMismatch" forceMatch={serverErrors.email}>
+    Please provide a valid email.
+  </Form.Message>
+</Form.Field>
+```
+
+## Open questions
+
+- Does this API make sense?
+  - Is it easy to use?
+  - Is everything named as you'd expect?
+- Does this API cover most form use-cases? Have we missed some use-cases?
```

Comment:
Q. How should I express different validations when one field is dependent on another?

For example, if I have a `country` field and an `age` field, as shown below, and I need the max/min value of age to be different depending on the value of the country field, how can I use it?

```jsx
<Form.Field name="country">
  <Form.Label>Country</Form.Label>
  <Form.Control asChild>
    <select>
      <option value="uk">United Kingdom</option>
      <option value="us">United States</option>
    </select>
  </Form.Control>
</Form.Field>

<Form.Field name="age">
  <Form.Label>Age</Form.Label>
  <Form.Control type="number" min="18" max="57" step="1" />
</Form.Field>
```

Reply:
you can use react state as you usually would:

```jsx
const [country, setCountry] = React.useState('');

<Form.Field name="country">
  <Form.Label>Country</Form.Label>
  <Form.Control asChild>
    <select value={country} onChange={event => setCountry(event.currentTarget.value)}>
      <option value="uk">United Kingdom</option>
      <option value="us">United States</option>
    </select>
  </Form.Control>
</Form.Field>

<Form.Field name="age">
  <Form.Label>Age</Form.Label>
  <Form.Control type="number" min={country === 'uk' ? 18 : 21} max="57" step="1" />
</Form.Field>
```

Reply:
Or you could also do it using a custom validation function as it has access to the entire form data so it can check other fields values.